### DISCUSSION TITLE
Native executable generation error while using maven native profile: Could not find target method: io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder.fromJDKProperties

### DISCUSSION THREAD

**Discussion by DiegoGutierrezFing** [quarkiverse/quarkus-cxf#1086] [Q&A]
Hi there, i hope you all are doing great.

I'm a quarkus begginer user coming from a Java EE (and some SpringBoot) background. After succesfully compiling, building and running a SOAP webservice application in classic JVM mode using the **quarkus-cxf** extension, i tried generating a **native mode** equivalent executable following the tutorials here:

https://quarkus.io/guides/building-native-image

As you might expect from a newcomer, I´m having some trouble i could not diagnose the underlying cause. This is what happens when compiling a native quarkus application using the native profile:

```shell
mvn clean package -Pnative
```

Which returns an error i don´t know how to interpret:

```shell
Error: Could not find target method: private static java.lang.String io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder.fromJDKProperties(java.lang.String,java.lang.String)
com.oracle.svm.core.util.UserError$UserException: Could not find target method: private static java.lang.String io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder.fromJDKProperties(java.lang.String,java.lang.String)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:73)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.findOriginalMethod(AnnotationSubstitutionProcessor.java:824)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleMethodInAliasClass(AnnotationSubstitutionProcessor.java:441)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleAliasClass(AnnotationSubstitutionProcessor.java:413)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:386)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:342)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.createAnnotationSubstitutionProcessor(NativeImageGenerator.java:966)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:868)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:575)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:535)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:403)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:580)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
```

 As far as i know, the class:

```java
io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder
```

 seems to be a dummy class whose method:

```java
io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder.fromJDKProperties(java.lang.String,java.lang.String)
```

only returns "null" :

```java
package io.quarkiverse.cxf.saaj.graal;

import com.oracle.svm.core.annotate.Substitute;
import com.oracle.svm.core.annotate.TargetClass;

@TargetClass(className = "jakarta.xml.soap.FactoryFinder")
final class Traget_javax_xml_soap_FactoryFinder {

    /**
     * The target method wants to read a properties file under {@code java.home} which does not work on GraalVM (there
     * is no
     * JRE distro at native runtime).
     *
     * @param factoryId
     * @param deprecatedFactoryId
     * @return
     */
    @Substitute
    private static String fromJDKProperties(String factoryId, String deprecatedFactoryId) {
        return null;
    }
}

public class SoapApiSubstitutions {
}
```

Can somebody give me a hint of what the error means, or if possible at all, how to fix it?

Also, in more general terms, Is it possible to generate native executables using the **quarkus.cxf** extension?

If not, is it in the roadmap?

Thanks a lot and greetings from Uruguay
Diego

Edited: corrected some grammar errors

**Comment by ppalaga** [comment]
> Also, in more general terms, Is it possible to generate native executables using the quarkus.cxf extension?

Yes, definitely. We state so in our docs https://docs.quarkiverse.io/quarkus-cxf/dev/user-guide/package-for-jvm-and-native.html

> `mvn clean package -Pnative`

That's right.

```
Could not find target method: private static java.lang.String io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder.fromJDKProperties(java.lang.String,java.lang.String)
```

`io.quarkiverse.cxf.saaj.graal.Traget_javax_xml_soap_FactoryFinder` is a GraalVM substitution - it replaces some method in native mode, in general such ones whose original code is known not to work in native. The error message seems to suggest that you have some version of `jakarta.xml.soap:jakarta.xml.soap-api` in your class path that does not contain the `jakarta.xml.soap.FactoryFinder.fromJDKProperties(String, String)` method. 

As a first step, make sure that you import Quarkus CXF and Quarkus BOMs as documented here: https://docs.quarkiverse.io/quarkus-cxf/dev/user-guide/create-project.html#dependency-management

If that does not help, please share your pom.xml.

**Reply by DiegoGutierrezFing** [reply]
Thanks a lot for the quick response!

I will check out the source code repository again as you suggested, but i'm sure that i manually included the artifact that includes the class with the method the error mentions with no avail:

```xml
<dependency>
      <groupId>io.quarkiverse.cxf</groupId>
      <artifactId>quarkus-cxf-saaj</artifactId>
</dependency>
```

The error message is the same regardless of whether I manually include that dependency or not.

I uploaded the pom.xml file, but in the meantime i will follow again the guide step by step as you mentioned.

[pom.txt](https://github.com/quarkiverse/quarkus-cxf/files/13241680/pom.txt)

**Reply by ppalaga** [reply]
Including `quarkus-cxf-saaj` really makes no difference. Your problem is probably a missing or broken dependencyManagement or a wrong version of `jakarta.xml.soap:jakarta.xml.soap-api`.

Could you paste the output of `mvn org.apache.maven.plugins:maven-dependency-plugin:3.6.1:tree` here?

**Reply by ppalaga** [reply]
> Could you paste the output of mvn org.apache.maven.plugins:maven-dependency-plugin:3.6.1:tree here?

Forget it, I noticed, you posted your pom.xml

**Reply by ppalaga** [reply]
Yeah, you have

```
    <dependency>
      <groupId>jakarta.xml.soap</groupId>
      <artifactId>jakarta.xml.soap-api</artifactId>
      <version>3.0.0</version>
    </dependency>
```

in your pom. It does not need to be there. 

I guess you also do not need any of those jakarta and sun.com artifacts - they or their equivalents are transitivelly pulled by quarkus-cxf.

Please try this: 

```
<?xml version="1.0"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>uy.com.antel.pi</groupId>
  <artifactId>tlk-acm-adapter-quarkus</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <properties>
    <compiler-plugin.version>3.11.0</compiler-plugin.version>
    <maven.compiler.release>17</maven.compiler.release>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
    <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
    <quarkus.platform.version>3.5.0</quarkus.platform.version>
    <skipITs>true</skipITs>
    <surefire-plugin.version>3.1.2</surefire-plugin.version>
  </properties>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>${quarkus.platform.artifact-id}</artifactId>
        <version>${quarkus.platform.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>quarkus-cxf-bom</artifactId>
        <version>${quarkus.platform.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-arc</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkiverse.cxf</groupId>
      <artifactId>quarkus-cxf</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-health</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkiverse.cxf</groupId>
      <artifactId>quarkus-cxf-saaj</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-junit5</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.apache.commons</groupId>
      <artifactId>commons-text</artifactId>
    </dependency>
  </dependencies>
  <build>
    <plugins>
      <plugin>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <version>${quarkus.platform.version}</version>
        <extensions>true</extensions>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
              <goal>generate-code</goal>
              <goal>generate-code-tests</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${compiler-plugin.version}</version>
        <configuration>
          <compilerArgs>
            <arg>-parameters</arg>
          </compilerArgs>
          <showDeprecation>true</showDeprecation>
        </configuration>
      </plugin>
      <plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>${surefire-plugin.version}</version>
        <configuration>
          <systemPropertyVariables>
            <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
            <maven.home>${maven.home}</maven.home>
          </systemPropertyVariables>
        </configuration>
      </plugin>
      <plugin>
        <artifactId>maven-failsafe-plugin</artifactId>
        <version>${surefire-plugin.version}</version>
        <executions>
          <execution>
            <goals>
              <goal>integration-test</goal>
              <goal>verify</goal>
            </goals>
            <configuration>
              <systemPropertyVariables>
                <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
                <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
                <maven.home>${maven.home}</maven.home>
              </systemPropertyVariables>
            </configuration>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
              <goal>generate-code</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>jaxb2-maven-plugin</artifactId>
        <version>3.1.0</version>
        <executions>
          <execution>
            <id>xjc</id>
            <goals>
              <goal>xjc</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <sources>
            <source>${basedir}/src/main/resources/xsd/</source>
          </sources>
          <clearOutputDir>false</clearOutputDir>
        </configuration>
      </plugin>
    </plugins>
  </build>
  <profiles>
    <profile>
      <id>native</id>
      <activation>
        <property>
          <name>native</name>
        </property>
      </activation>
      <properties>
        <skipITs>false</skipITs>
        <quarkus.package.type>native</quarkus.package.type>
      </properties>
    </profile>
  </profiles>
</project>
```

**Reply by DiegoGutierrezFing** [reply]
Thanks a lot Mr. Palaga!

As you suggested, i changed the pom.xml file deleting the dependency:

```xml
<dependency>
      <groupId>jakarta.xml.soap</groupId>
      <artifactId>jakarta.xml.soap-api</artifactId>
      <version>3.0.0</version>
</dependency>
```

And now i can generate a native image executable without any problems.

By the way, about your observation:

> I guess you also do not need any of those jakarta and sun.com artifacts - they or their equivalents are transitivelly pulled by quarkus-cxf.

The sun artifacts you are probably refering to are related to some packages needed to add ws-addressing capabilities (i.e adding ws-addressing headers) to the soap clients generated by cxf as per this javaee metro guide:

[adding-soap-headers-when-sending-requests](https://javaee.github.io/metro/doc/user-guide/user-guide.html#adding-soap-headers-when-sending-requests)

I didn´t knew if them were necessary or not with quarkus: in the SpringBoot original project, they were needed to be declared explicitly in the pom.xml file or the project would refuse to build. Surely i need to check those out and also try to improve my maven knowledge.

PS: This app was a "port" of another one i made with the same functionality using SpringBoot. **Diagnosing the original issue with quarkus native image generation was a real head scratcher for me for several days, so again thanks a lot for your help**. The generated executable seems to run well on my workstation, but until next monday i could not test it in a pre-production environment, because i can't access the SOAP endpoints outside my job.

**Comment by ppalaga** [comment]
> By the way, about your observation:
> 
> > I guess you also do not need any of those jakarta and sun.com artifacts - they or their equivalents are transitivelly pulled by quarkus-cxf.
> 
> The sun artifacts you are probably refering to are related to some packages needed to add ws-addressing capabilities (i.e adding ws-addressing headers) to the soap clients generated by cxf as per this javaee metro guide:
> 
> [adding-soap-headers-when-sending-requests](https://javaee.github.io/metro/doc/user-guide/user-guide.html#adding-soap-headers-when-sending-requests)
> 
> I didn´t knew if them were necessary or not with quarkus: in the SpringBoot original project, they were needed to be declared explicitly in the pom.xml file or the project would refuse to build. Surely i need to check those out and also try to improve my maven knowledge.

Metro and CXF are both implementations of JAX-WS. Quarkus CXF focuses solely on porting CXF to Quarkus and Metro is fully out of our scope. This code might be what you want for setting addressing headers without relaying on Metro: 

https://github.com/quarkiverse/quarkus-cxf/blob/2.6.0/integration-tests/ws-addressing/src/test/java/io/quarkiverse/cxf/it/ws/addressing/server/WsAddressingDecoupledTest.java#L53-L85
